src/expr/attribute_internals.h

```diff
@@ -145,16 +149,372 @@ inline uint64_t GetBitSet(uint64_t bit)
+                           [](const value_type& a, const value_type& b) {
+                             return a.first < b.first;
+                           });
+      if (range.first != range.second)
+      {
+        // key already present, don't insert
+        return std::make_pair(iterator{}, false);
+      }
+
+      return std::make_pair(d_contents.insert(range.first, std::move(p)), true);
+    }
+
+    const_iterator find(uint64_t key) const
+    {
+      auto range =
+          std::equal_range(d_contents.begin(),
+                           d_contents.end(),
+                           std::make_pair(key, V{}),
+                           [](const value_type& a, const value_type& b) {
+                             return a.first < b.first;
+                           });
+      if (range.first == range.second)
+        // not in map
+        return d_contents.end();
+      else
+        return range.first;
+    }
+
+    iterator find(uint64_t key)
+    {
+      auto range =
+          std::equal_range(d_contents.begin(),
+                           d_contents.end(),
+                           std::make_pair(key, V{}),
+                           [](const value_type& a, const value_type& b) {
+                             return a.first < b.first;
+                           });
+      if (range.first == range.second)
+        return d_contents.end();
+      else
+        return range.first;
+    }
+
+    iterator erase(iterator pos) { return d_contents.erase(pos); }
+
+    V& operator[](uint64_t key)
+    {
+      iterator it =
+          std::lower_bound(d_contents.begin(),
+                           d_contents.end(),
+                           std::make_pair(key, V{}),
+                           [](const value_type& a, const value_type& b) {
+                             return a.first < b.first;
+                           });
+      if ((it == d_contents.end()) || (it->first != key))
+      {
+        // not in map
+        it = d_contents.insert(it, std::make_pair(key, V{}));
+      }
+      return (*it).second;
+    }
+
+    // range insert
+    template <typename Iter>
+    void insert(Iter beg, Iter end)
+    {
+      for (Iter it = beg; it != end; ++it)
+      {
+        iterator found_it =
+            std::lower_bound(d_contents.begin(),
+                             d_contents.end(),
+                             it->first,
+                             [](const value_type& a, const value_type& b) {
+                               return a.first < b.first;
+                             });
+        if ((found_it != d_contents.end()) && (it->first == found_it->first))
+        {
+          // this key is already present in the map. replace it:
+          found_it->second = it->second;
+        }
+        else
+        {
+          d_contents.insert(found_it, *it);
+        }
+      }
+    }
+
+   private:
+    Container d_contents;
+  };
+
+  template <typename Parent, typename L1It, typename L2It>
+  class Iterator
```

Comment:
It would be good to add some comments for this iterator class to describe what it actually does since this iterates over the top-level std::unordered_map and the attribute vector.